pkg/common-controller/snapshot_controller.go

```diff
@@ -671,6 +672,16 @@ func (ctrl *csiSnapshotCommonController) createSnapshotContent(snapshot *crdv1.V
 		},
 	}
 
+	nodeName, err := ctrl.getManagedByNode(snapshot)
+	if err != nil {
+		return nil, err
+	}
+	if nodeName != "" {
```

Comment:
Why not simply do
```go
if node := os.Getenv("NODE_NAME"); node != "" {
  snapshotContent.Labels = map[string]string{
    utils.VolumeSnapshotContentManagedByLabel: nodeName,
  }
}
```
Since we know the side car is running on this node, and we know the snapshot content will reside on that particular node. I see no real reason to go through looking up the PV, then looking up all the nodes, and then matching the affinity of the PV to a node. When in the end we will just return NODE_NAME anyway.

Reply:
Also are we sure there are no labels on the snapshotContent already? We are overriding anything that is there (if anything)

Reply:
For your 1st question- the common snapshot_controller will still run as a single instance and not on every node. This controller is responsible for creating VolumeSnapshotContent, so it will figure out the right node and will label the VolumeSnapshotContent instance with that node.
The csi-snapshotter, on the other hand, will run as a daemon set on each node and will filter out the VolumeSnapshotContent objects by matching the node label.

For you 2nd question, we are adding a new label here, so is it possible that we could be overriding some other label?

Reply:
Oh I totally missed this is the common controller, and not the side car. I figured I would optimize this and we wouldn't have to do all the lookups of the nodes and volumes. One more question on line 646 of the same file right before creating the snapshotcontent object this is called:
```go
class, volume, contentName, snapshotterSecretRef, err := ctrl.getCreateSnapshotInput(snapshot)
```
volume is the PV that we are looking up, could we pass the volume to the function so we can avoid having to look it up again? Secondly why are we looking up nodes to match names against. Regardless of the status of the node (down/up/gone) we would want to label the snapshot content with the node the original volume was on. So why not just look at the affinity of the PV, and return the node name from it. That way we don't have to look up the nodes.

For the second question, I just looked at the code above again, and the snapshotContent was just created a few lines above and it has no labels. So we are good.

Reply:
Yeah, I suppose I could reuse the volume that we retrieve on line 646. I'll make the changes for that. Thanks for pointing it out.

For the next question, we are looking up nodes and then matching is against the node affinity and not simply getting the label from the volume and applying is to VolumeSnapshotContent is, because I don't think there's a global label key that all the drivers use to manage node affinity. Each driver will have its own topology keys, I suppose. So I wasn't sure what should be the label key which would give us the node name as value. Let me know if you can think of any better way to handle this.

Reply:
You make a very good point about the topology key being different for each driver. For my driver I have been working on the reverse problem (restoring a snapshot into a volume). And I can make assumptions about the topology key in there and avoid the node lookup.

My main concern is the following (unlikely) scenario. The PV was created, we are about to look up the nodes, and the node we are interested went down, and is no longer in the list. So now when we loop over the nodes, we will never find the node that matches and thus we won't be able to label the snapshot content. And I would very much like to be able to count on the label being there, for the restore of a snapshot.

One problem I have not been able to solve is efficiently locate the node the snapshot is on if I want to restore it on another node (This can happen if the k8s scheduler schedules a pod that uses a WFFC PVC that has a datasource of a snapshot, on a different node than where the snapshot resides). If the label exists on the snapshot content, it would be trivial to locate the node.

I cannot think of a better mechanism than what you have done at this point, so I think I am happy right now with the code. I can just fail the restore if the node label is not there or empty.